docs/architecture/xamarin.forms/virtualized-collections.md

```diff
@@ -0,0 +1,264 @@
+```
+_Note the use of `'itemMarker` to enforce the type of widget items_
+
+Keeping the original items allows us to directly compare them on each update to determine if we need to update the UI.
+```fs
+(fun (a, b) -> ScalarAttributeComparers.equalityCompare(a.OriginalItems, b.OriginalItems))
+```
+
+Currently, scalar attributes in Fabulous have 2 generic parameters: the `'inputType` and `'modelType`.
+
+The `'inputType` is what we expect the users to provide us.
+The `'modelType` is an optimized representation of the same data (eg. `'inputType` is `int list`, `'modelType` will be `int[]`).
+
+Before storing the attribute value, we convert `'inputType` to `'modelType` through the `Convert` function declared in the `ScalarAttributeDefinition`; `'inputType` is never stored.
+
+But the comparison function shown just above is only done between 2 `'modelType`s, but ListView/CollectionView expect an `IEnumerable` and not a `WidgetItems`.
+
+So to support this, we need another generic parameter `'valueType` and the corresponding `ConvertValue: 'modelType -> 'valueType` function.
+```fs
+type ScalarAttributeDefinition<'inputType, 'modelType, 'valueType> =
+    { (...)
+      Compare: 'modelType -> 'modelType
+      ConvertValue: 'modelType -> 'valueType }
+```
+
+With this, we can still store and compare `WidgetItems`, but when we actually need to apply the value to the XF property, we call `ConvertValue` to transform it. Attributes that don't need conversion can use the `id` function to make it transparent.
+
+For `WidgetItems`, we build an IEnumerable on the fly using the original items and the template function before assigning it to the XF property `ItemsSource`.
+```fs
+(fun modelValue -> seq { for x in modelValue.OriginalItems do modelValue.Template x })
+```
+
+Now Xamarin.Forms has a list of Widgets, and thanks to `seq`, it will only enumerate the items it needs to display.
+
+#### Loading Widgets into rows
+
+Unlike DataTemplate in MVVM apps, instead of setting bindings to capture the raw data inside `BindingContext`, we now have a `Widget`.
+
+To make it work, we need 2 things:
+- A DataTemplateSelector that will know which DataTemplate to create based on the widget
+- A DataTemplate that will listen to `BindingContextChanged` and run the Reconciler when a widget is attached to the row
+
+```fs
+// IsHeader is only for grouping
+type WidgetDataTemplateSelector internal (node: IViewNode, isHeader: bool, getWidget: obj -> Widget) =
+    inherit DataTemplateSelector()
+    
+    /// Reuse data template for already known widget target type
+    let cache = Dictionary<Type, WidgetDataTemplate>()
+    
+    override _.OnSelectTemplate(item, _) =
+        let widget = getWidget item
+        let widgetDefinition = WidgetDefinitionStore.get widget.Key
+        let targetType = widgetDefinition.GetTargetType(widget)
+        match cache.TryGetValue(targetType) with
```

Comment:
So that is the place that calls to `GetTargetType`, so caching is needed here because `GetTargetType` is going to be expensive for `Memo specifically`.

Hm, I think that will work but I wonder if there is an alternative solution. I think conceptually prohibiting `Memo` inside collection view is fine. (you usually don't want to do that regardless), but I'm not sure how to express that

Reply:
Actually, it's already prohibited 😅 
We can just change `GetTargetType` to a simple field (all widgets know their type at creation), and set a default `Unchecked.defaultOf<_>` for Memo since it won't be called anyway.

![Screenshot 2022-01-13 at 09 43 43](https://user-images.githubusercontent.com/6429007/149296036-93d27ea9-639a-44e2-8bab-1e07c0f528ee.png)

Reply:
YES!!!!